[RFC] net: vertexcom: mse102x: Fix SPI interrupt handling

lategoodbye opened the pull request:
This series fixes the handling of the SPI interrupt in the MSE102x driver. The current solution is consider as a request for comments.

Additional it improves the related debug capabilites by introducing an optional DT property to use the interrupt line as GPIO (currently not intended for upstream).

lategoodbye committed:
net: vertexcom: mse102x: Implement GPIO interrupt support

It's hard to debug the SPI interrupt level, so implement the possibility
to allocate the SPI interrupt as a GPIO. This allows to debug the current
level with the common tools.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
arm64: dts: imx93-charge-som-dc-evb: Switch MSE102x to GPIO interrupt

This allows to debug the SPI interrupt level.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: drop invalid cmd stats

Since the SPI implementation on the MSE102x MCU is in software,
it cannot reply to SPI commands in busy state. So drop the scaring
statistics about "invalid" command replies.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: Fix possible stuck of SPI interrupt

The MSE102x doesn't provide any SPI commands for interrupt handling.
So in case the interrupt fired before the driver requests the IRQ,
the interrupt will never fire again. In order to fix this always poll
for pending packets after opening the interface.

Fixes: 2f207cbf0dd4 ("net: vertexcom: Add MSE102x SPI support")
Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: Fix SPI IRQ type

According to the MSE102x documentation the trigger type is a
high level.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
dt-bindings: vertexcom-mse102x: Fix IRQ type

According to the MSE102x documentation the trigger type is a
high level.

Fixes: 2717566f6661 ("dt-bindings: net: add Vertexcom MSE102x support")
Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: Drop retry from mse102x_rx_pkt_spi

Since mse102x_rx_pkt_spi is also used for polling and the SPI IRQ
type is level based, there is actually no need for a retry
mechanism within mse102x_rx_pkt_spi. So drop it and simplify
the receive path.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: Add warning about IRQ trigger type

The example of the initial DT binding of the Vertexcom MSE 102x suggested
a IRQ_TYPE_EDGE_RISING, which is wrong. So warn everyone to fix their
device tree.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: Add range check for CMD_RTS

Since there is no protection in the SPI protocol against electrical
interferences, the driver shouldn't blindly trust the length payload
of CMD_RTS. So do at least a lower bounds check for incoming frames.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

mhei commented on drivers/net/ethernet/vertexcom/mse102x.c:
```diff
@@ -308,27 +311,33 @@ static void mse102x_rx_pkt_spi(struct mse102x_net *mse)
-	if ((cmd_resp & CMD_MASK) != CMD_RTS) {
+	if ((cmd_resp & CMD_MASK) == CMD_RTS) {
+		rxlen = cmd_resp & LEN_MASK;
+		if (rxlen < ETH_ZLEN) {
```
Now that we know that the rxlen should be max 1518, I'd add a check here too using the defines ETH_FRAME_LEN + VLAN_HLEN

mhei commented on drivers/net/ethernet/vertexcom/mse102x.c:
```diff
@@ -308,27 +311,33 @@ static void mse102x_rx_pkt_spi(struct mse102x_net *mse)
+	 * So assume the maximum Ethernet frame length.
+	 */
+	if (drop)
+		rxlen = 1518;
```
here too, would use ETH_FRAME_LEN + VLAN_HLEN

lategoodbye committed:
net: vertexcom: mse102x: Fix RX error handling

In case the CMD_RTS got corrupted by interferences, the MSE102x
doesn't allow a retransmission of the command. Instead the Ethernet
frame must be shifted of the SPI FIFO. Since the actual length is
unknown, assume the maximum possible value.

Fixes: 2717566f6661 ("dt-bindings: net: add Vertexcom MSE102x support")
Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>

lategoodbye committed:
net: vertexcom: mse102x: Extend range check for CMD_RTS

Introduce a upper bounds check for incoming frames in order to
catch invalid CMD_RTS.

Signed-off-by: Stefan Wahren <stefan.wahren@chargebyte.com>
